### ISSUE TITLE
System.IO.IOException on yahb.CopyModule.createDirectoryList()

### ISSUE THREAD

**Issue by FloSchwalm** [asdfjkl/yahb#16]
There is a problem with recent Adobe Acrobat Licensing Service that makes yahb and other backup software break when trying to access the log folders of the software for backup purposes. This has been reported to Adobe and will hopefully be worked on (https://community.adobe.com/t5/illustrator-discussions/com-adobe-dunamis-folder-cannot-be-backed-up-by-backup-solutions/td-p/14559757)

However, it would probably be possible to fix this (or work around it) in yahb as well in case this happens in the future with this or other software.

Here is the output from yahb just before the crash
```
Unbehandelte Ausnahme: System.IO.IOException: Der Prozess kann nicht auf die Datei "C:\Users\flo\AppData\Roaming\com.adobe.dunamis\d225650b-9c1f-4738-97e3-94e805951049\v1\0" zugreifen, da sie von einem anderen Prozess verwendet wird.
   bei System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   bei System.IO.FileSystemEnumerableIterator`1.CommonInit()
   bei System.IO.FileSystemEnumerableIterator`1..ctor(String path, String originalUserPath, String searchPattern, SearchOption searchOption, SearchResultHandler`1 resultHandler, Boolean checkHost)
   bei System.IO.Directory.EnumerateDirectories(String path)
   bei yahb.CopyModule.createDirectoryList()
   bei yahb.Program.Main(String[] args)
```

Whatever Acrobat Licensing Service does here, there is currently no catch for `IOException` in `createDirectoryList()`. So maybe add this and keep working on the remaining directories as is already done for some other exceptions.

~~Maybe if I get this right we could also try a fix with different EnumerationOptions.~~
_Update: Turns out after getting to the point in the .NET code where the exception is thrown we do not have an option to prevent this with different `EnumerationOptions`. I'll leave the analysis here anyway in case someone wants to follow the thought process._ 

yahb calls `EnumerateDirectories()` with only one argument, the directory to start from
https://github.com/asdfjkl/yahb/blob/96435995c00eb58db1625b82a1d62224768df949/yahb/CopyModule.cs#L75

This means that the unused parameters are filled with default options, the `EnumerationOptions` being set to `EnumerationOptions.Compatible`
https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs#L216

Notably `EnumerationOptions.Compatible` means that `IgnoreInaccessible` is set to `false`
https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/IO/EnumerationOptions.cs#L20-L21

We thus end up calling `EnumerateDirectories()` with all available parameters internally
https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs#L223-L224

Which then calls `InternalEnumeratePaths()` defined just above the different `EnumerateDirectories()` definitions
https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs#L196-L214

This leads to another internal call to `FileSystemEnumerableFactory.UserDirectories()` defined here
https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEnumerableFactory.cs#L128-L140

Creating a new `FileSystemEnumerable` instance defined here
https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEnumerable.cs#L14-L38

And here is the interesting part, at the end of the constructor we create a `DelegateEnumerator` which according to the source code comment ensures that we get possible IO exceptions for the target directory right at the beginning
https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEnumerable.cs#L35-L37

This `DelegateEnumerator` creates a `FileSystemEnumerator`
https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEnumerable.cs#L60-L68

Ath the end of the `FileSystemEnumerator` constructor it calls its method `Init()`
https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEnumerator.cs#L31-L43

Which is implemented in the Windows specific file and creates a directory handle to check for any IO exceptions
https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEnumerator.Windows.cs#L48-L50

So unfortunately there is no try/catch here and changing the `EnumerationOptions` won't help, we just have to catch the IOException in yahb.

Maybe custom `EnumerationOptions` would help at locations where `IgnoreInaccessible` is actually used?
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEnumerator.Windows.cs#L115

I think with `IgnoreInaccessible` and `RecurseSubdirectories` set to `true` in `EnumerationOptions` we may get the full directory list with much less going back and forth between yahb's `createDirectoryList()` and the .NET functions.

But certainly not related to this bug then.

**Comment by asdfjkl** [comment]
thanks for reporting!

sigh ... just submitted a bugfix version to winget, as there was an uncaught, exception, now there is another one :-)

Did a quick recompile, catching IOException. Would highly appreciate if you could test whether this one runs through without crashing... unfortunately I do not have Adobe software here for testing.

[yahb.zip](https://github.com/asdfjkl/yahb/files/15049865/yahb.zip)

**Comment by FloSchwalm** [comment]
LGTM

**Comment by asdfjkl** [comment]
fixed in latest release (1.0.9.0)
